### ISSUE TITLE
Fails to install or run in Windows 10 Pro bld 1809 (x64)

### ISSUE THREAD

**Issue by MartyMacGyver** [sqlitebrowser/sqlitebrowser#1771]
Details for the issue
--------------------

#### What did you do?

After updating from Windows 10 Pro 1803 to 1809 yesterday, my system notified me my old 3.0.8 sqlitebrowser was disabled (and in fact it was removed during the upgrade). So, I grabbed the latest version 3.11.1 (x64) and tried installing it.

#### What did you expect to see?

I expected it to at least install.

#### What did you see instead?

When the options had been selected and the install began, it failed saying it couldn't be installed. If I had more info (like a log) I would share it, but it doesn't appear the install creates one.

#### How'd you manage to work around this vexing issue?

I've never seen Windows do this before, so I was perplexed by it. I don't care if night mode works, so long as I can run the app. Interestingly, it suggested a reboot would be required when I attempted the install (but it never got that far).

I tried the slightly older portable version and it seemed to work. Then I used [LessMSI](https://github.com/activescott/lessmsi) to manually extract the contents of `DB.Browser.for.SQLite-3.11.1-win64.msi`. Within it, I found two folders: `DB Browser for SQLite` and `System64`. I'm not sure what the contents of `System64` are necessary for (is this what the installer warned I'd need to reboot for?) but simply running `DB Browser for SQLite\DB Browser for SQLite.exe` worked fine (I just created a quick DB to try it out). I moved that folder where I want to run this from and so far, so good.

*So, the purpose of this ticket is to let you know I may not be the only person running into this, and that the very latest regular channel Windows 10 seems unwilling to install this normally.* There's a slightly hacky solution to it (manually extracting the msi, etc.), and the System64 files strike me as odd - I'm pretty sure these (or later versions thereof) are fully redundant and shouldn't be getting installed at all on Windows 10:

https://docs.microsoft.com/en-us/uwp/win32-and-com/win32-apis

Useful extra information
-------------------------

The info below often helps, please fill it out if you're able to. :)

#### What operating system are you using?

- [x] Windows: ( _version: 10 Pro 1809 fully up-to-date, using night mode )
- [ ] Linux: ( _distro:_ ___ )
- [ ] Mac OS: ( _version:_ ___ )
- [ ] Other: ___

#### What is your DB4S version?

- [x] 3.11.1
- [ ] 3.11.0
- [ ] 3.10.1
- [ ] Other: ___

#### Did you also

- [x] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [x] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by justinclift** [comment]
> So, the purpose of this ticket is to let you know I may not be the only person running into this, and that the very latest regular channel Windows 10 seems unwilling to install this normally. 

Thanks, it's useful info.  Hopefully we'll be able to figure out what the problem is and get it solved.  Our team and wider Community have some people with fairly decent Windows skills (I'm not one of them :wink:), so they'll probably have suggestions for things to investigate over the next day or two.

Some general thoughts, in case they help... :wink:

1. There are two downloads for Win64, a .msi and a .zip.  The .zip is just [an extracted .msi](https://github.com/sqlitebrowser/sqlitebrowser/blob/e52f2054f2aa1d0bbee8dbeac635e74963022f94/installer/windows/nightly_build_scripts/win64build.bat#L102), created using the msiexec command line tool instead of LessMSI.  It sounds like the .exe's in the .zip should work fine, but would you be ok to give them a go to confirm?

2. It's probably worth checking if right clicking on the downloaded .msi in Windows Explorer, then choosing properties shows anything interesting.  eg maybe there's an "Unblock" button in there or similar?  I've seen it with Win7 development VM's, and people have mentioned seeing similar with Win10 too.  It's one of the first things to check, and you've probably already done it... but making sure the basics are taken care of is generally useful. :smile:

**Comment by MartyMacGyver** [comment]
The fact that during the upgrade to build 1809 Windows very specifically removed this because of some incompatibility was really unusual.

Ticking "Unblock" didn't help either, and it still says it'll reboot (I suspect because it's apparently trying to put files in the system area, though *why* I don't know). I've attached the visual output - same as what I first saw:

![0fd7383e-2a33-41ce-97c3-d3fe996df38b png](https://user-images.githubusercontent.com/1078647/53549454-09570680-3ae9-11e9-9f75-753e79859812.jpg)

The zip file (how'd I miss that??) does indeed work fine.

I was able to log the install process via:

    msiexec /l*v "log.log" /i "DB.Browser.for.SQLite-3.11.1-win64.msi"`

The log is quite verbose and has a fair amount of personalized info in it. The notable parts:

At the point when the installer mentions rebooting, the log says, `RESTART MANAGER: Did detect that a critical application holds file[s] in use, so a reboot will be necessary.` It doesn't say what files it's updating though (and why?)

Then it tries to uninstall the old NSIS-installed version (long gone now) `CA_NSIS_UNINSTALL`. This appears to end up failing:

```
MSI (s) (A8:C8) [23:45:56:617]: Note: 1: 1721 2: CA_NSIS_UNINSTALL 3: C:\bin\DB Browser for SQLite\Uninstall.exe 4: /S 
MSI (s) (A8:C8) [23:45:56:617]: Note: 1: 2205 2:  3: Error 
MSI (s) (A8:C8) [23:45:56:617]: Note: 1: 2228 2:  3: Error 4: SELECT `Message` FROM `Error` WHERE `Error` = 1721 
Error 1721. There is a problem with this Windows Installer package. A program required for this install to complete could not be run. Contact your support personnel or package vendor. Action: CA_NSIS_UNINSTALL, location: C:\bin\DB Browser for SQLite\Uninstall.exe, command: /S 
MSI (s) (A8:C8) [23:45:57:462]: Note: 1: 2205 2:  3: Error 
MSI (s) (A8:C8) [23:45:57:462]: Note: 1: 2228 2:  3: Error 4: SELECT `Message` FROM `Error` WHERE `Error` = 1709 
MSI (s) (A8:C8) [23:45:57:462]: Product: DB Browser for SQLite -- Error 1721. There is a problem with this Windows Installer package. A program required for this install to complete could not be run. Contact your support personnel or package vendor. Action: CA_NSIS_UNINSTALL, location: C:\bin\DB Browser for SQLite\Uninstall.exe, command: /S 

Action ended 23:45:57: InstallFinalize. Return value 3.
```

That appears to be the main failure.

[Looking at the source](https://github.com/sqlitebrowser/sqlitebrowser/blob/master/installer/windows/product.wxs), it seems this is a relatively recent change - for grins, I decided just now to install the older `DB.Browser.for.SQLite-3.10.1-win64.exe` and then uninstall it and then install this to see what it does.

*[Time passed...]*

It all worked this time, though it still requires a reboot (even though the app already works). Not at all sure why this new installer is copying things over existing system files (`C:\Windows\system32` evidently) when it's not really necessary. That bothers me, and I think I'll uninstall the MSI and stick with the no-install version instead for now.

**Comment by justinclift** [comment]
Excellent, that looks like some good in-depth debugging happening right there. :smile:

It does seem like the NSIS uninstaller isn't quite happy for some reason.  That'll probably be something @karim would be interested in, as time permits.

> Not at all sure why this new installer is copying things over existing system files (C:\Windows\system32 evidently) when it's not really necessary.

My understanding (hopefully not too far wrong!) is this is from the installer generator ([Wix Toolset](http://wixtoolset.org)), and should only be copying things over when they're not present on the system.  Maybe also when they're outdated too (unsure).  I am **not** any kind of expert with this stuff. :smile:

**Comment by justinclift** [comment]
Oh... wonder if the new installer is somehow leaving registry keys from the prior 3.10.1 release around, thereby fooling itself into thinking 3.10.1 is still present?

That would kind of explain the behaviour.  And if that's the case, may we need to add in some extra clean-up of the old reg keys, and also add some extra checking.  eg maybe verify the uninstaller path given actually has a file there

**Comment by justinclift** [comment]
Off topic, but...

>  SELECT `Message` FROM `Error` WHERE `Error` = 1721 

Wonder if that's an embedded SQLite? :smile:

**Comment by chrisjlocke** [comment]
Good grief.  This ticket reads like an episode after my Mum has been on her laptop for half an hour.  "It came up with this message, so I deleted everything, then uninstalled everything.  Now what do I do?  Do I click the blue button?  I've clicked the blue button.  Now it has another error.  Do I do anything?  Well I clicked that and now its asked me to install Linux, so where do I do that?"

The log is mentioning "C:\bin\DB Browser for SQLite\Uninstall.exe".  Have you got that installed?  By default, the MSI will want to install in the normal place of "c:\program files".  It appears you've installed various copies of DB4S all over the place.  

So, lets clean up the mess.  From Control Panel, uninstall all previous copies of DB4S.  Old 3.8's, alpha 3.10s, etc.  Decide where you want DB4S - Program files, or c:\bin.  I presume c:\bin was from a portable copy?

If you run the MSI, it should run through fine now.  Because you had older copies, but couldn't uninstall them, it was having issues.

DB4S doesn't install much (its pretty much self-contained) so if an uninstall doesn't work, you can just delete the directories, run the uninstall from Control Panel again, and (usually) WIndows is sensible enough to say, "The program you're trying to uninstall can't be found - remove it from the Add/Remove programs list?".  This is enough to 'clean' up after itself.

As a last resort, you don't need to install MSI extractors, unpackers or hacks.  On the downloads page there is a .ZIP where you can extract and run DB4S.  You can even unpack this to 'c:\program files\DB Browser for SQLite' if you want.

So just to confirm.  Open COntrol Panel, go to 'Programs' or click 'Uninstall a Program'

![image](https://user-images.githubusercontent.com/7872218/53555817-99716c00-3b3a-11e9-839f-9caec8f4f677.png)

DB4S should be in the list under 'DB Browser for SQLite'

![image](https://user-images.githubusercontent.com/7872218/53555876-b4dc7700-3b3a-11e9-8cf1-b6f9561c7191.png)

**Comment by chrisjlocke** [comment]
> The fact that during the upgrade to build 1809 Windows very specifically removed this because of some incompatibility was really unusual.

3.8 is fairly old, and I assume it used an even older version of the C++ runtime.  Win 10 by default has its own newer versions, so there is a slight conflict there.  If you'd installed 3.8 again, Win 10 would have probably added some compatibility layer hoohicky so that it all would work, but can't do that 'after the event' so to speak...

**Comment by karim** [comment]
Hello @MartyMacGyver and thanks for taking the time to report and debug this issue.

I will try to explain all of the issues you had here, so excuse the long post. :smiley: 

> After updating from Windows 10 Pro 1803 to 1809 yesterday, my system notified me my old 3.0.8 sqlitebrowser was disabled (and in fact it was removed during the upgrade).

I have never seen anything, but I'm on Windows 7, so I don't know if it is something Microsoft introduced in Windows 10. Doing a quick search, did not return anything, but I found this, and I don't know if it is related...

- https://www.cnet.com/news/windows-10-october-2018-update-may-delete-personal-files-during-some-upgrades/

The post mentions deleting personal files, but it doesn't talk about disabling apps or removing them, so I'm not sure.

> If I had more info (like a log) I would share it, but it doesn't appear the install creates one.

The installer creates a log automatically during installation. This was requested a few months back (https://github.com/sqlitebrowser/sqlitebrowser/issues/1545#issuecomment-425623047).

>  Then I used [LessMSI](https://github.com/activescott/lessmsi) to manually extract the contents of `DB.Browser.for.SQLite-3.11.1-win64.msi`.

Like Justin said, you don't have to do this manually anymore. Just download the .zip file which is simply the extracted .msi file.

If you have Visual Studio installed, there is a bundled application, called Orca, that will let you do that.

> Within it, I found two folders: `DB Browser for SQLite` and `System64`. I'm not sure what the contents of `System64` are necessary for (is this what the installer warned I'd need to reboot for?)

It is just the [Visual C++ Redistributable Package](https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads). It is required by DB Browser for SQLite.

We bundle it with the installer because we don't know if the user has it installed or not. If it is installed already, then it is skipped (and no reboot is required), if not then it is installed (and you must reboot).

If you are using the MSI, these files are installed in the system folder and a reboot is required because Windows can't replace system files without a reboot.

> but simply running `DB Browser for SQLite\DB Browser for SQLite.exe` worked fine (I just created a quick DB to try it out). I moved that folder where I want to run this from and so far, so good.

An alternative way is to copy these files inside the application folder. Windows will automatically load these DLLs from there when they are found, and if not it will try looking for them in the system folder. That's how DLL loading in Windows works.

If you are using the .zip file then you will find these files already inside the application folder, so no need to install them or reboot.

We do exactly what you just did. We extract the .msi file, and we get two folders, one for our application, and the other for VCRedist, we move the DLLs inside the application folder, then just .zip it all.

And why there are two folders when you extracted the .msi file? I guess because how the .msi is built, as two different modules.

> and the System64 files strike me as odd - I'm pretty sure these (or later versions thereof) are fully redundant and shouldn't be getting installed at all on Windows 10

We really have no control over this. :smiley: 

We simply include the VCRedist, as a module, within the .msi and your OS should decide what to do with it.

Recently, the [VCRedist was split](https://devblogs.microsoft.com/cppblog/the-great-c-runtime-crt-refactoring/) into two parts. The [Universal CRT](https://devblogs.microsoft.com/cppblog/introducing-the-universal-crt/), which you should have on an already updated Windows 10, since it is part of the operating system now, and the Visual C++ libraries, which I think are the one being installed because they are missing. They get updated with every new version of Visual Studio we use to compile the application.

> Then it tries to uninstall the old NSIS-installed version (long gone now) `CA_NSIS_UNINSTALL`. This appears to end up failing

`CA_NSIS_UNINSTALL` is a CustomAction (hence the CA :smiley:) to look for older versions of DB4S and remove it before installing the new version. It is only meant to uninstall version made with NSIS.

It detect older versions by looking through the registry to find where you installed it, since it could be anywhere on your system. Then run the uninstall command silently.

> MSI (s) (A8:C8) [23:45:56:617]: Note: 1: 1721 2: CA_NSIS_UNINSTALL 3: C:\bin\DB Browser for SQLite\Uninstall.exe 4: /S

`C:\bin\` is where you had the older version, and that's where it was looking. But since the files are gone now without actually being uninstalled normally, to remove the registry entries, the MSI still thinks you have an older version and tries to uninstall it first.

>  I decided just now to install the older `DB.Browser.for.SQLite-3.10.1-win64.exe` and then uninstall it and then install this to see what it does.

Nice trick. :smiley: 

Installing yet another NSIS version simply recreated the registry entry to point it to v3.10.1 location, and uninstalling it removed the registry entry. Now the MSI doesn't fail anymore. Even if you kept the 3.10.1 installed the installer would have found it and uninstalled it without a problem.

> Not at all sure why this new installer is copying things over existing system files (`C:\Windows\system32` evidently) when it's not really necessary.

That shouldn't happen. My guess is, multiple restarts were needed cause each version (3.0.8, 3.10.1, 3.11.1) is shipped with different version of VCRedist, and each one is required to be installed. Also, VCRedist is installed as a different app (with an entry in Control Panel) and when you uninstall the app these VCRedist stays on your system, just in case another app depends on it.

Newer versions of DB4S does not need that anymore. No VCRedist in Control Panel, and it gets cleanly uninstalled if you no longer need it.

> That bothers me, and I think I'll uninstall the MSI and stick with the no-install version instead for now.

We understand, sorry for that. But again, that shouldn't happen, and it is better to find where the problem is.

Can you try removing all installed versions, and then install one of the nightlies (not the last one) so you get VCRedist installed as well, and probably restart when required, then after restart install another nightly over it (a newer one) and see if it will ask you to reboot again?

It shouldn't ask for reboot this time since you have the same version of VCRedist installed already. If it asks for a reboot, then this is a bug.

**Comment by karim** [comment]
@justinclift 

> Wonder if that's an embedded SQLite? :smile:

Yes, it is actually. :smile: 

![orca](https://user-images.githubusercontent.com/8761081/53562113-2a574000-3b59-11e9-8e30-c5dade80d4d3.jpg)

The .msi file is just a database file. The Windows Installer (msiexec.exe) will read this database and do actions based on the tables on it.

It is strange when you think about it, but at the time it was made, it was something new. :smiley:

**Comment by justinclift** [comment]
> Yes, it is actually.

Heh Heh Heh, that's pretty cool. :smile:

Also, could be handy for debugging stuff at times too.

Good explanation above btw, completely makes sense. :smile:

**Comment by MartyMacGyver** [comment]
@chrisjlocke I was being verbose as it was an unusual issue, and quite exact. I've got more than a little experience with the platform and did not post this spuriously - I posted it because it highlighted an unusual situation related to the old installer and as 1809 continues to roll out it might affect other users.

I previously had 3.10.1 installed (not 3.8, not 3.10 alpha), and as I said Windows posted a notification that it specifically removed that and only that application during the upgrade for incompatibility reasons.

I already explained above that I reinstalled 3.10.1, uninstalled it (because I suspect however windows uninstalled it automatically, it was the wrong way) and then *successfully* installed 3.11.1 via the msi installer. So, the main issue is resolved.

Why did Windows uninstall 3.10.1 (installed using the msi installer it came with) during the 1803 to 1809 upgrade? Why then did it refuse to allow 3.11.1 to be installed? Those were the open questions that led to my thinking it would be good to let you know this happened.

**Comment by justinclift** [comment]
> Those were the open questions that led to my thinking it would be good to let you know this happened.

Thank you, they're the right things.  Don't worry about @chrisjlocke's comment, he's normally on point, but everyone has the occasional bad day. :wink:

**Comment by MartyMacGyver** [comment]
To be fair, his story about his Mum was hilarious (and I've been on that side of the system way too often)... It just had absolutely no relevance here. If I posted "I install v0.1 ten years ago and bought a Mac today and I can't open my Windows" then yeah, I'm probably not on the way to success.

Anyway, hope it all at least ends up helping someone else.

**Comment by justinclift** [comment]
> To be fair, his story about his Mum was hilarious...

Yep. :smile:
